plugins/arrow-flight-rpc/src/main/java/org/opensearch/arrow/flight/bootstrap/FlightClientManager.java

```diff
@@ -0,0 +1,265 @@
+ * This class maintains a pool of Flight clients for internode communication,
+ * handles client lifecycle, and responds to cluster state changes.
+ *
+ * <p>The manager implements ClusterStateListener to automatically update
+ * client connections when nodes join or leave the cluster. </p>
+ */
+public class FlightClientManager implements ClusterStateListener, AutoCloseable {
+    private static final Version MIN_SUPPORTED_VERSION = Version.V_3_0_0;
+    private static final Logger logger = LogManager.getLogger(FlightClientManager.class);
+    static final int LOCATION_TIMEOUT_MS = 1000;
+    private final ExecutorService grpcExecutor;
+    private final ClientConfiguration clientConfig;
+    private final Map<String, ClientHolder> flightClients = new ConcurrentHashMap<>();
+    private final Client client;
+
+    /**
+     * Creates a new FlightClientManager instance.
+     *
+     * @param allocator          Supplier for buffer allocation
+     * @param clusterService     Service for cluster state management
+     * @param sslContextProvider Provider for SSL/TLS context configuration
+     * @param elg                Event loop group for network operations
+     * @param threadPool         Thread pool for executing tasks asynchronously
+     * @param client             OpenSearch client
+     */
+    public FlightClientManager(
+        BufferAllocator allocator,
+        ClusterService clusterService,
+        @Nullable SslContextProvider sslContextProvider,
+        EventLoopGroup elg,
+        ThreadPool threadPool,
+        Client client
+    ) {
+        grpcExecutor = threadPool.executor(ServerConfig.FLIGHT_CLIENT_THREAD_POOL_NAME);
+        this.clientConfig = new ClientConfiguration(
+            Objects.requireNonNull(allocator, "BufferAllocator cannot be null"),
+            Objects.requireNonNull(clusterService, "ClusterService cannot be null"),
+            sslContextProvider,
+            Objects.requireNonNull(elg, "EventLoopGroup cannot be null"),
+            Objects.requireNonNull(grpcExecutor, "ExecutorService cannot be null")
+        );
+        this.client = Objects.requireNonNull(client, "Client cannot be null");
+        clusterService.addListener(this);
+    }
+
+    /**
+     * Returns a Flight client for a given node ID.
+     *
+     * @param nodeId The ID of the node for which to retrieve the Flight client
+     * @return An OpenSearchFlightClient instance for the specified node
+     */
+    public FlightClient getFlightClient(String nodeId) {
+        ClientHolder clientHolder = flightClients.getOrDefault(nodeId, null);
+        return clientHolder != null ? clientHolder.flightClient : null;
+    }
+
+    /**
+     * Returns the location of a Flight client for a given node ID.
+     *
+     * @param nodeId The ID of the node for which to retrieve the location
+     * @return The Location of the Flight client for the specified node
+     */
+    public Location getFlightClientLocation(String nodeId) {
```

Comment:
We don't need this method, it is only used in tests

Reply:
this is needed in the next PR - https://github.com/opensearch-project/OpenSearch/pull/17313

Reply:
removed for now